design/crd-status.md

```diff
@@ -0,0 +1,316 @@
+
+#### IPAddressPool Status
+
+This should be easy enough, as it will require extending the current `IPAddressPool`
+CRD with a `Status` section:
+
+```yaml
+apiVersion: metallb.io/v1beta1
+kind: IPAddressPool
+metadata:
+  name: example
+  namespace: metallb-system
+spec:
+  addresses:
+  - 192.168.10.0/24
+  - 192.168.9.1-192.168.9.5
+  - fc00:f853:0ccd:e799::/124
+status:
+  availableIPV4: 45
+  availableIPV6: 145
+  assignedIPV4: 5
+  assignedIPV6: 52
+```
+
+```go
+    type IPAddressPoolStatus struct {
+        assignedIPV4  int
+        assignedIPV6  int
+        availableIPV4 int
+        availableIPV6 int
+}
+```
+
+#### Configuration Status
+
+Given that the configuration is composed by multiple CRs, there are few cases
+where a given configuration is invalid because of a single CR (i.e. invalid IP formatting).
+
+The majority of the scenarios involve multiple CRs which are not compatible together.
+For this reason, we think a global `ConfigurationStatus` indicator is better and
+easier to understand, compared to a "per resource" status that tells if the resource
+is valid or not.
+
+```yaml
+apiVersion: metallb.io/v1beta1
+kind: ConfigurationStatus
+metadata:
+  name: config-status
+  namespace: metallb-system
+status:
+    validConfig: false
+    error: "peer 1.2.3.4 has myAsn different from 1.2.3.5, in FRR mode all myAsn must be equal"
+```
+
+```go
+    type MetalLBConfigurationStatus struct {
+        validConfig bool
+        lastError   string
+    }
+
+    type ConfigurationStatus struct {
+        metav1.TypeMeta   `json:",inline"`
+        metav1.ObjectMeta `json:"metadata,omitempty"`
+        Status            MetalLBConfigurationStatus `json:"status,omitempty"`
+    }
+```
+
+**Note:** given the fact that the configuration is parsed both by the speakers and
+by the controller, we might want to expose the status of each component. However,
+there is currently no logic in the configuration parsing that depends on the component.
+For this reason, we can have an initial status that is produced only by the controller
+(which runs in single instance) to validate the status.
+
+As an alternative, we might consider having a state per component, named after the
+single component that produces the status, but in general having a single place to
+check seems more straightforward.
+
+If we go with the _per component_ scenario, we might add a `loadedConfiguration`
+field that exposes the latest loaded configuration by that component. This can't
+be done if we let the controller to expose the single configuration because
+what's loaded might depend on the order the CRs are received with.
+
+#### BGP / BFD Session state
+
+Because of the scalability concerns expressed above, the idea is to produce an instance
+of the resource _per peer / node_, which exposes the state of the session between
+the speaker running on a given node and a given peer.
+
+The name of a given instance will be like `nodename-peer`, and each instance will
```

Comment:
I understand the scalability concerns. However, I believe that an instance per node and peer pair is too much. Instead, we can have once instance per node. Taking as an example the number of nodes you mentioned above and 2 peers, there will be 1000 instances instead of 2000.

Reply:
yeah, that might work. My guess is we'll have a high number of nodes and a manageable number of peers.
This means that when a session (or two) flickers, we'll move the status related to the node, but the size should be manageable.

Moreover, we'll control how often we publish this, so it should be ok. @oribon thoughts?

Reply:
Not sure I follow, do you suggest here having a resource per node, and there specifying all of its peers + connection status?
If so wouldn't it (over)complicate querying how a given peer behaves with the nodes of the cluster compared to what suggested here (is it something we'd like to be simple?)
I'm not against it, but I feel that the current way "talks" better with the rest of the proposal

If you think the number of objects here play an important role then it sounds reasonable to condense them in a one-per-node object

Reply:
> Wouldn't it (over)complicate querying how a given peer behaves with the nodes of the cluster compared to what suggested here (is it something we'd like to be simple?) I'm not against it, but I feel that the current way "talks" better with the rest of the proposal
> 

Right, I did not remember this part :sweat_smile: 

> If you think the number of objects here play an important role then it sounds reasonable to condense them in a one-per-node object

I think both versions would be fine performance wise especially considering that we would sample the status anyway and that we would not update anything if the session keeps being stable.

This is probably (until now) the remaining pending discussion. Given both are more or less the same, but the original one would favor navigability, I am inclined to tend for that. Basically what we'd loose with the "all the peers for one node" is a way to say "give me the status of all the sessions for the peer xy".

@mlguerrero12 do you have reasons to say "an instance per node per peer" is too much? Is the concern about the number of objects?

Reply:
yes, my concern is the number of objects and also from the user perspective, I think it is nicer to see in one place all sessions of a speaker instead of having to get each one of them separately

Reply:
from the user's perspective it won't change. Because of the labels, they will still have a way to look at all the session from a given speaker selecting the right label.

The difference is, in the all-in-one it won't be possible to catch all the session for a given peer, and the amount of data moved when a session is flickering will be slightly more.
All in all, I feel that having one item representing a single session is more natural. The considerations made on endpointslices vs endpoints ([here](https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/#motivation)) seem to tend to the fact that it's better to have smaller objects rather than a big one, so the number should not be a technical issue.

Reply:
ok